File: packages/dotnet-sdk/src/TurboDocx/HttpClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace TurboDocx;

/// <summary>
/// HTTP client for TurboDocx API requests
/// </summary>
internal class HttpClient : IDisposable
{
    private readonly System.Net.Http.HttpClient _client;
    private readonly TurboDocxClientConfig _config;
    private readonly string _baseUrl;
    private bool _disposed;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    public HttpClient(TurboDocxClientConfig config)
    {
        _config = config;
        _baseUrl = config.BaseUrl ?? "https://api.turbodocx.com";
        _client = new System.Net.Http.HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };
    }

    private void SetHeaders(HttpRequestMessage request, bool includeContentType = true)
    {
        if (includeContentType)
        {
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }

        if (!string.IsNullOrEmpty(_config.AccessToken))
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _config.AccessToken);
        }
        else if (!string.IsNullOrEmpty(_config.ApiKey))
        {
            request.Headers.Add("X-API-Key", _config.ApiKey);
        }
    }

    private async Task HandleErrorResponse(HttpResponseMessage response)
    {
        var content = await response.Content.ReadAsStringAsync();
        string message = $"HTTP {(int)response.StatusCode}: {response.ReasonPhrase}";
        string? code = null;

        try
        {
            var errorData = JsonSerializer.Deserialize<ErrorResponse>(content, JsonOptions);
            if (errorData != null)
            {
                message = errorData.Message ?? errorData.Error ?? message;
                code = errorData.Code;
            }
        }
        catch
        {
            // Use default message if parsing fails
        }

        throw new TurboDocxException(message, (int)response.StatusCode, code);
    }

    public async Task<T> GetAsync<T>(string path, CancellationToken cancellationToken = default)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, _baseUrl + path);
        SetHeaders(request);

        var response = await _client.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            await HandleErrorResponse(response);
        }

        var result = await response.Content.ReadFromJsonAsync<T>(JsonOptions, cancellationToken);
        return result!;
    }

    public async Task<byte[]> GetRawAsync(string path, CancellationToken cancellationToken = default)
    {
        var request = new HttpRequestMessage(HttpMethod.Get, _baseUrl + path);
        SetHeaders(request, includeContentType: false);

        var response = await _client.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            await HandleErrorResponse(response);
        }

        return await response.Content.ReadAsByteArrayAsync(cancellationToken);
    }

    public async Task<T> PostAsync<T>(string path, object? data, CancellationToken cancellationToken = default)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, _baseUrl + path);
        SetHeaders(request);

        if (data != null)
        {
            var json = JsonSerializer.Serialize(data, JsonOptions);
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        var response = await _client.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            await HandleErrorResponse(response);
        }

        var result = await response.Content.ReadFromJsonAsync<T>(JsonOptions, cancellationToken);
        return result!;
    }

    public async Task<T> UploadFileAsync<T>(
        string path,
        byte[] file,
        string fileName,
        Dictionary<string, string>? additionalData,
        CancellationToken cancellationToken = default)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, _baseUrl + path);
        SetHeaders(request, includeContentType: false);

        var content = new MultipartFormDataContent();
        content.Add(new ByteArrayContent(file), "file", fileName);

        if (additionalData != null)
        {
            foreach (var (key, value) in additionalData)
            {
                content.Add(new StringContent(value), key);
            }
        }

        request.Content = content;

        var response = await _client.SendAsync(request, cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            await HandleErrorResponse(response);
        }

        var result = await response.Content.ReadFromJsonAsync<T>(JsonOptions, cancellationToken);
        return result!;
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed)
        {
            if (disposing)
            {
                _client.Dispose();
            }
            _disposed = true;
        }
    }

    private class ErrorResponse
    {
        public string? Message { get; set; }
        public string? Error { get; set; }
        public string? Code { get; set; }
    }
}

File: packages/ruby-sdk/lib/turbodocx/http_client.rb
# frozen_string_literal: true

require "faraday"
require "faraday/multipart"
require "json"

module TurboDocx
  class HttpClient
    DEFAULT_BASE_URL = "https://api.turbodocx.com"

    def initialize(api_key:, access_token: nil, base_url: nil)
      @api_key = api_key
      @access_token = access_token
      @base_url = (base_url || DEFAULT_BASE_URL).chomp("/")
    end

    def get(path)
      response = connection.get(path)
      handle_response(response)
    end

    def get_raw(path)
      response = connection.get(path)
      handle_raw_response(response)
    end

    def post(path, body)
      response = connection.post(path) do |req|
        req.headers["Content-Type"] = "application/json"
        req.body = body.to_json
      end
      handle_response(response)
    end

    def upload_file(path, file, file_name, form_data)
      response = multipart_connection.post(path) do |req|
        req.body = form_data.merge(
          file: Faraday::Multipart::FilePart.new(
            StringIO.new(file),
            "application/octet-stream",
            file_name
          )
        )
      end
      handle_response(response)
    end

    private

    def connection
      @connection ||= Faraday.new(url: @base_url) do |f|
        f.request :json
        f.response :json
        f.adapter Faraday.default_adapter
        f.headers.merge!(auth_headers)
      end
    end

    def multipart_connection
      @multipart_connection ||= Faraday.new(url: @base_url) do |f|
        f.request :multipart
        f.response :json
        f.adapter Faraday.default_adapter
        f.headers.merge!(auth_headers)
      end
    end

    def auth_headers
      headers = {}
      headers["X-API-Key"] = @api_key if @api_key
      headers["Authorization"] = "Bearer #{@access_token}" if @access_token
      headers
    end

    def handle_response(response)
      return response.body if response.success?

      handle_error(response)
    end

    def handle_raw_response(response)
      return response.body if response.success?

      handle_error(response)
    end

    def handle_error(response)
      body = response.body || {}
      message = body["message"] || "API Error"
      code = body["code"]

      raise Error.new(message, status_code: response.status, code: code)
    end
  end
end
